### PR TITLE
 Remove declarations from rootcling

### PR DESCRIPTION
**Pull request by yamaguchi1024**
Remove the declaration of TClingRuntime.h, string.h, stddef.h, and stdlib.h. Also delete TClingRuntime.h as it was the only use.

### CODE REVIEW

**Comment by vgvassilev** [comment]
@pcanal, the change seems to work with roottest, shall we go forward with it?

**Comment by pcanal** [comment]
@vgvassilev for compleness sake, on which platform did you run the full roottest suite?

**Comment by vgvassilev** [comment]
@phsft-bot build just on ubuntu16/native with flags -Dctest_test_exclude_none=on

**Comment by vgvassilev** [comment]
@yamaguchi1024, the failure seems related to the removal of `#include "TObject.h"`.

File: core/dictgen/src/rootcling_impl.cxx [hunk 1]
```diff
@@ -4506,14 +4506,8 @@ int RootClingMain(int argc,
       // rootcling
       if (interp.declare("namespace std {} using namespace std;") != cling::Interpreter::kSuccess
             // CINT uses to define a few header implicitly, we need to do it explicitly.
-            || interp.declare("#include <assert.h>\n"
-                              "#include <stdlib.h>\n"
-                              "#include <stddef.h>\n"
-                              "#include <string.h>\n"
-                             ) != cling::Interpreter::kSuccess
             || interp.declare("#include \"Rtypes.h\"\n"
-                              "#include \"TClingRuntime.h\"\n"
```
Comment (vgvassilev) [inline core/dictgen/src/rootcling_impl.cxx]: We can kill `TClingRuntime.h` as this is its only use.

File: core/metacling/src/TCling.cxx [hunk 2]
```diff
@@ -1369,17 +1369,12 @@ TCling::TCling(const char *name, const char *title, const char* const argv[])
    if (fromRootCling) {
       fInterpreter->declare("#include \"RtypesCore.h\"\n"
                             "#include <string>\n"
-                            "using std::string;\n"
-                            "#include <cassert>\n");
```
Comment (pcanal) [inline core/metacling/src/TCling.cxx]: This changes may or may not good/appropriate **but** definitively do not fall under the commit's description.   Can you put them in a separate commit with an explanation ...  thanks.

**Comment by yamaguchi1024** [comment]
@phsft-bot build just on slc6/gcc62 with flags -Druntime_cxxmodules=On -Dctest_test_exclude_none=on

**Comment by pcanal** [comment]
> Remove redundant declaration from TCling

Could you explicit where the other include are?  Thanks.

**Comment by yamaguchi1024** [comment]
@phsft-bot build

**Comment by yamaguchi1024** [comment]
@phsft-bot build

**Comment by vgvassilev** [comment]
@pcanal, can you greenlight this PR?

**Comment by pcanal** [comment]
@phsft-bot build flags -Dctest_test_exclude_none=on

**Comment by pcanal** [comment]
looks good, let's check on the full roottest suite.

**Comment by vgvassilev** [comment]
@phsft-bot build with flags -Dctest_test_exclude_none=on

**Comment by vgvassilev** [comment]
@phsft-bot build with flags -Dctest_test_exclude_none=on

**Comment by vgvassilev** [comment]
@phsft-bot build with flags -Dctest_test_exclude_none=on

**Comment by pcanal** [comment]
The error on MacOS could possibly be related to the header files removed.  Re-running the test to verify that it is not from something else than this PR.

@phsft-bot build with flags -Dctest_test_exclude_none=on

**Comment by vgvassilev** [comment]
@phsft-bot build with flags -Dctest_test_exclude_none=on

**Comment by vgvassilev** [comment]
@phsft-bot build!

**Comment by vgvassilev** [comment]
@phsft-bot build with flags -Dctest_test_exclude_none=on

**Comment by yamaguchi1024** [comment]
@phsft-bot build

**Comment by Axel-Naumann** [comment]
The "good reason for those includes to stay" is that people currently have code like
```
void foo(const string& str) {
  if (strstr(str.c_str(), "abc") {}
}
```
and that will fail to parse after your change. I.e. we're breaking users' code.

**Comment by yamaguchi1024** [comment]
I see, I'm not sure if string is in PCH and maybe includes can stay for PCH, but at least for modules we know that we're loading stl and libc at startup time, no?

**Comment by Axel-Naumann** [comment]
As we discussed, I prefer to have ROOT's code work with or without PCH or PCM. In general, let's use PCH and PCM as an optimization, but let's keep ROOT's code valid (and backward compatible for users) without:

- rootcling needs to be exposed to types as before, in case people (maybe unknowingly) rely on e.g. `<string>` being included and `using namespace std` being injected (`#pragma link C++ class MyClass<string>`). OTOH, `string.h` is likely not needed in rootcling's runtime.
- the interpreter-version of `ClassDef` should assume that the interpreter has not seed `Rtypes.h` before, for consistency.

Thanks!

**Comment by Axel-Naumann** [comment]
@phsft-bot build!

File: core/dictgen/src/rootcling_impl.cxx (hunk 1 above)
Comment (pcanal) [reply core/dictgen/src/rootcling_impl.cxx]: Was the "why remove the other includes ... ? We should not rely on transitive inclusions through TObject.h."
resolved?

Comment (Axel-Naumann) [reply core/dictgen/src/rootcling_impl.cxx]: I believe so - these were ROOT headers, and they are now not part of the change anymore.

Comment (pcanal) [reply core/dictgen/src/rootcling_impl.cxx]: I am confused now ... I  thought this was talking about:
``` 
                            "#include <stdlib.h>\n"
                              "#include <stddef.h>\n"
                              "#include <string.h>\n"
```
which were the only header in TClingRunTime.h (+ stdio.h and math.h)

Comment (Axel-Naumann) [reply core/dictgen/src/rootcling_impl.cxx]: They were targeted at an earlier version of this PR that removed `TObject.h` and `Rtypes.h`. See also my comment below

> rootcling needs to be exposed to types as before, in case people (maybe unknowingly) rely on e.g. <string> being included and using namespace std being injected (#pragma link C++ class MyClass<string>). OTOH, string.h is likely not needed in rootcling's runtime.
